### DISCUSSION TITLE
Intellisense documentation not entirely working when sourcing/running other notebooks

### DISCUSSION THREAD

**Discussion by mwh3780** [microsoft/pylance-release#2650] [Enhancement]
### Discussed in https://github.com/microsoft/pylance-release/discussions/2037

<div type='discussions-op-text'>

<sup>Originally posted by **mwh3780** November  4, 2021</sup>
I recently inherited a process where code is spread out and stored in various Jupyter notebooks and they're linked together by calling each other through `%run`ning / sourcing each other.

In trying to use my normal Jupyter setup in VS Code with Pylance as the language server I've noticed that Intellisense doesn't completely work. For objects and functions defined in notebooks that are sourced as well as any modules imported in these notebook, autocompletion does work when typing the names of these objects/functions/modules, but the hover over documentation doesn't. When I hover over any of these objects, whether they are custom objects, functions, methods or modules that are defined in notebooks that are sourced the hover over documentation simply shows `Object name: Any`. And just for reference, JupyterLab and classic Jupyter Notebook don't have this issue.

I tried looking for help or troubleshooting guides for this particular issue, but didn't find anything that helped.

Since reproducing the behavior involves different files, below is some minimal code that shows the behavior:

```
### Code stored in file named random_code.ipynb
import pandas as pd
import numpy as np

df_1 = pd.DataFrame(np.random.randint(0, 10, size = (3, 3)), columns = ["A", "B", "C"])

def fun1(x, y = 1):
    '''
    Some function

    Arguments:
    ...

    Examples:
    ...
    '''
    return(x + y)
```

```
%run "/Users/mwh3780/Box/DataExplorer/Test/hammer_sandbox/Sub folder/random_code.ipynb"

# Hover over documentation does not work for any of the following :
df_1.sort_values()
fun1(1, 2)
pd.DataFrame()
```

Sees the object in the environment
<img width="579" alt="image" src="https://user-images.githubusercontent.com/23129139/140347672-0b82f0bc-869c-4259-9ab7-d73407c50495.png">

But documentation doesn't work
<img width="404" alt="image" src="https://user-images.githubusercontent.com/23129139/140347043-d273e060-f51d-40b4-bba6-65a1d87e5cdc.png">

<img width="191" alt="image" src="https://user-images.githubusercontent.com/23129139/140356667-ab151ca4-c680-4e19-b9a8-2bb9d77c8c81.png">

<img width="185" alt="image" src="https://user-images.githubusercontent.com/23129139/140356778-95e6776c-8662-46aa-a8d7-1391ad0b7f83.png">

I'm using:
OS: macOS Big Sur 11.4
VS Code: Version: 1.61.2 (Universal)
Jupyter Extension Version: v2021.9.1101343141
Pylance Extension Version: v2021.11.0
Python Extension Version: v2021.10.1365161279</div>

**Comment by judej** [comment]
The IntelliSense is not working since we don't support magic commands right now.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Reply by Latvala** [reply]
It would be very nice if this is going to be fixed :). I'm using two notebooks for data analysis, one is front-end and one is backend. I have built this model that other team members who are not so familiar with coding, can reuse my back-end modules.

**Comment by DavidHJong** [comment]
It would be awesome if IntelliSense could better handle variables and functions from %run commands in Jupyter notebooks. Here are a few ideas:

**Static Analysis:**

- Parse the called notebook for imports, global variables, and functions.

**Heuristic Analysis:**

- Use heuristics to detect things like return values from functions.

**Integration:**

- Work with the Python extension to provide accurate hover-over info and linting.

This would make navigating and understanding code so much easier. Thanks for considering!
